[clang-tidy] Add readability-use-span-first-last check (#118074)
llvm/llvm-project: clang-tools-extra/docs/clang-tidy/checks/readability/use-span-first-last.rst

.. title:: clang-tidy - readability-use-span-first-last

readability-use-span-first-last
===============================

Checks for uses of ``std::span::subspan()`` that can be replaced with clearer
``first()`` or ``last()`` member functions. These dedicated methods were added 
to C++20 to provide more expressive alternatives to common subspan operations.

Covered scenarios:

> ==================================== ==================================

Reviewer: Columns width.

Author: Done.

.. title:: clang-tidy - readability-use-span-first-last

readability-use-span-first-last
===============================

> Checks for uses of ``std::span::subspan()`` that can be replaced with clearer

Reviewer: Please synchronize first statement with Release Notes.

instead of equivalent ``subspan()``. These dedicated methods were added to C++20 
to provide more expressive alternatives to common subspan operations.

Covered scenarios:

> =========================== ============

Reviewer: Second column should be expanded on two characters to fit `s.first(n)` with all back-ticks.

Author: thank you, super sorry that this is such a back n forth, but my brain seems to be wired a way to not being able to figure RST out 🙈

Reviewer: You just need to look on diff :-)

> =============================== ============

Reviewer: Please reduce first column on 4 characters and add two characters to second column, so all cell content will be covered.

Author: resolved.

> ========================== ====================

Reviewer: Please enlarge first column to `s.subspan(s.size() - n)` (with all back-ticks) width. Second column is too wide.

Author: rst formatting is really tough 😬

Author: Done.

> =============================== ================

Author: Done.

------------------------- --------------------
``s.subspan(0, n)``       ``s.first(n)``
``s.subspan(s.size() - n)`` ``s.last(n)``
========================== ====================

Non-zero offset with count (like ``subspan(1, n)``) or offset-only calls 
(like ``subspan(n)``) have no clearer equivalent using ``first()`` or 
``last()``, so these cases are not transformed.

> This check is only active when C++20 or later is used.

Reviewer 2: Missing newline

> This check is only active when C++20 or later is used.

Reviewer: ```suggestion
This check is enabled for C++20 or later.
```
